components/script/dom/htmlimageelement.rs

```diff
@@ -356,20 +390,244 @@ impl HTMLImageElement {
     }
 
     /// <https://html.spec.whatwg.org/multipage/#update-the-source-set>
-    fn update_source_set(&self) -> Vec<DOMString> {
+    fn update_source_set(&self) {
+        // Step 1
+        *self.source_set.borrow_mut() = HTMLImageElement::new_source_set();
+
+        // Step 2
         let elem = self.upcast::<Element>();
-        // TODO: follow the algorithm
-        let src = elem.get_string_attribute(&local_name!("src"));
-        if src.is_empty() {
-            return vec![]
+        let parent = elem.upcast::<Node>().GetParentElement();
+        let is_parent_picture = match parent.clone() {
+            Some(xparent) => xparent.is::<HTMLPictureElement>(),
+            None => false
+        };
+        let nodes = parent.unwrap().upcast::<Node>().ChildNodes();
+        let elements = match is_parent_picture {
+            true =>
+            nodes.iter().filter_map(DomRoot::downcast::<Element>)
+                        .map(|n| DomRoot::from_ref(&*n)).collect(),
+            false => vec![DomRoot::from_ref(&*elem)],
+        };
+
+        // Step 3
+        let _width = match elem.get_attribute(&ns!(), &local_name!("width")) {
+            Some(x) => { match parse_length(&x.value()) {
```

Comment:
Let's fix up the indentation here - we can either remove the { following the => (and the corresponding }), or we can move the match to the next line and indent everything under it by an extra four spaces.